packages/flutter_tools/lib/src/flutter_manifest/parse_result.dart

```diff
@@ -0,0 +1,40 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+final class ParseResult<T> {
```

Comment:
I realize that I probably should write another code sample that mirrors my first one but instead using a similar abstraction (e.g. just a tuple). Then we could compare those two

Reply:
_Actually_, I think the best thing to do here is say YAGNI and close this. I come back to this issue after I the asset-transformer parsing code lands.

Reply:
> Consider the situation where a parsing function (like `AssetsEntry.parse`) needs to parse multiple things and then combine the results.
> 
> ```dart
> // In AssetsEntry.parse, we can (and should?) parse transformers and flavors separately so that
> // we can report all issues found, not just any issues found in the first parsing failure.
> final ParseResult<List<String>> flavorsParseResult = parseListNullable<String>(
>   yaml[_flavorKey],
>   'flavors list of assets entry "$path"',
>   'String',
> );
> 
> final ParseResult<List<AssetTransformerEntry>> transformersParseResult =
>     _parseTransformersSection(yaml[_transformersKey], path);
> 
> final List<String> errors = <String>[
>   ...switch(flavorsParseResult) {
>     ValueParseResult<List<String>?>() => <String>[],
>     ErrorParseResult<List<String>>(errors: final List<String> e) => e,
>   },
>   ...switch(transformersParseResult) {
>     ValueParseResult<List<AssetTransformerEntry>>() => <String>[],
>     ErrorParseResult<List<AssetTransformerEntry>>(errors: final List<String> e) => e,
>   },
> ];
> 
> if (errors.isNotEmpty) {
>   return ErrorParseResult<AssetsEntry>(errors);
> }
> 
> final AssetsEntry entry = AssetsEntry(
>   uri: uri,
>   // We could have something like ParseResult::asValue that does this for
>   // us (and throws a cleaner exception that a cast error), but the
>   // fundamental idea would be the same.
>   flavors: Set<String>.from((flavorsParseResult as ValueParseResult<List<String>>).value),
>   transformers: (transformersParseResult as ValueParseResult<List<AssetTransformerEntry>>).value,
> );
> 
> return ValueParseResult<AssetsEntry>(entry);
> ```

I would handle it with a switch-case statement (not expression) where we assign whatever we're wrapping to an appropriate local, like:

```dart
  final ParseResult<List<String>> flavorsParseResult =
      parseListNullable<String>(
    yaml[_flavorKey],
    'flavors list of assets entry "$path"',
    'String',
  );

  final ParseResult<List<AssetTransformerEntry>> transformersParseResult =
      _parseTransformersSection(yaml[_transformersKey], path);

  List<String>? flavors;
  List<AssetTransformerEntry>? transformers;
  List<String> errors = <String>[];
  switch (flavorsParseResult) {
    case ValueParseResult<List<String>>(value: final v):
      flavors = v;
    case ErrorParseResult<List<String>>(errors: final e):
      errors.addAll(e);
  }

  switch (transformersParseResult) {
    case ValueParseResult<List<AssetTransformerEntry>>(value: final v):
      transformers = v;
    case ErrorParseResult<List<AssetTransformerEntry>>(errors: final e):
      errors.addAll(e);
  }

  if (errors.isNotEmpty) {
    return ErrorParseResult<AssetsEntry>(errors);
  }

  final AssetsEntry entry = AssetsEntry(
    uri: uri,
    flavors: flavors!.toSet(),
    transformers: transformers,
  );

  return ValueParseResult<AssetsEntry>(entry);
```

but punting on this until later SGTM.